### PR TITLE
Add lazylist equal predicate

### PR DESCRIPTION
**Pull request by mars0i**
This PR adds the function `equal` to `LazyList`.  It's a response to issue #809.

The function definition is essentially the one suggested by kantian on [discuss.ocaml.org](https://discuss.ocaml.org/t/equality-of-lazylists/1192/11); it's a slight modification of the existing definition of `for_all2`.

(The first six commits have nothing to do with equal; I mistakenly caused my copy of master to get out of sync, and I had to fix it.  Only the last three commits are relevant.)

### CODE REVIEW

**Comment by gasche** [comment]
I like the patch, but not the shape of the patchset which is very noisy. You could use the method I proposed in https://github.com/ocaml-batteries-team/batteries-included/pull/810#issuecomment-348159639 to just cherry-pick the three last commits in a new branch starting from an up-to-date master.

**Comment by mars0i** [comment]
Github now allows "squashing" multiple commits into a single commit.  It looks like this has to done in the main repo, or enabled there, at least  https://github.com/blog/2141-squash-your-commits

File: src/batLazyList.ml [hunk 1]
```diff
@@ -619,6 +619,14 @@ let for_all2 p l1 l2 =
...
   in aux l1 l2
 
+let equal eq l1 l2 =
+  let rec aux l1 l2 =
+    match (next l1, next l2) with
+    | (Cons (h1, t1), Cons(h2, t2)) -> eq h1 h2 && (aux t1 t2)
+    | (Nil, Nil)                    -> true
+    | (Cons _, Nil) | (Nil, Cons _) -> false
+  in aux l1 l2
+
```
Comment (gasche) [inline src/batLazyList.ml]: As all new functions, this must come with unit tests. For the syntax of our inline tests, see the tests on `append` in the same module.

**Comment by UnixJunkie** [comment]
Thanks for your contribution.

**Comment by UnixJunkie** [comment]
@mars0i I updated the changelog file for you, please don't forget next time.
With more follow through, you could have had this merged way earlier.
Maybe just pinging in here after having added unit tests would have been enough.

**Pull request merged**